Misc fixes to get the CLI integration test more reliable

thallgren opened the pull request:
This PR contains a number of minor bug fixes to Telepresence, discovered while running the
`pkg/client/cli/telepresence_test.go` tests, and also a cleanup of that test to get more consistent
handling of the config.

I recommend reviewing one commit at a time.

Commit:
Remove unnecessary concurrency control around darwin DNS resolver files.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that msdDSNResponderHelper isn't killed on Big Sur.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix race condition when shutting down connector.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
General cleanup of config.yml timeout settings for telepresence cli test

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that the SetSearchPath method never hangs

Insert a small buffer for search path events, and if multiple entries
are in that buffer, disregard all entries but the last.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add debug logging to track all gRPC calls to the connector

Adds a `dlog.Debug(c, "called")` and `dlog.Debug(c, "returned")` to
all calls to the connector's gRPC API.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't use kubectl rollout status -w <workload> on k3s

See https://github.com/rancher/rancher/issues/21324 for more info on
this change.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung commented on pkg/client/cli/telepresence_test.go:
```diff
@@ -1834,3 +1761,55 @@ func telepresenceContext(ctx context.Context, args ...string) (string, string) {
+	config.LogLevels.RootDaemon = logrus.DebugLevel
+
+	to := &config.Timeouts
+	to.PrivateAgentInstall = 240 * time.Second
```
If we have to bump all these timeouts in the tests, that makes me think maybe our default timeouts are too short?  While we need to overwrite config to test certain things, and some we need to overwrite for all tests (e.g. the registry), I think we should try to use as many defaults as possible in our tests so we are testing, as best as possible, what people get out of the box when they install Telepresence.  So perhaps bump the timeouts in `config.go` and remove setting the defaults here

thallgren replied:
Not sure I agree. Our situation, where we use a sometimes extremely slow CI and not very fast cluster is, IMO out of the ordinary and this is a better safe than sorry timeout tailored for that use-case. For instance, bumping this default to 240s will leave the user hanging for four full minutes until telepresence gives up. That's quite exceptional and it's very likely to make the user unhappy when the problems are elsewhere (such as failure to pull the image, etc).

I also think that our approach of telling the user what times out on very slow systems is sufficient. It gives the user a chance to trim timeouts for a slow system without having a negative impact on all others.

donnyyung replied:
yeah fair enough. Maybe it's worth trying to use more performant clusters in our CI in the future. My main concern is wanting to make sure our timeouts are good on average because we don't want a user's first use of telepresence to result in a timeout since that's not the best UX imo. But that's not exactly a problem for this PR and that should probably be a metric's based decision and not one based on a "hunch" which is all I have at this point.

thallgren replied:
Totally agree with that. We should collect metrics for how long various things take. Perhaps we do already?

donnyyung replied:
I don't think we do, but it should be easy enough to sent a report every time we timeout w/ what the timeout was and what its value was.. so we can track that :) I'll make an issue just so we don't lose track of it.

donnyyung commented on pkg/client/daemon/dns/flush_darwin.go:
```diff
@@ -2,15 +2,58 @@ package dns
+	dlog.Debug(c, "Flushing DNS")
+	if majorProductVersion >= 11 {
+		// Big Sur, a killall -HUP mDNSResponder doesn't restart the mDNSResponderHelper but killall -TERM of both, does`
+		dlog.Debug(c, "Flushing DNS")
```
Either change to "Flushing DNS >= Big Sur" or delete this one since we have the same log at line 48.

donnyyung commented on pkg/client/daemon/dns/flush_darwin.go:
```diff
@@ -2,15 +2,58 @@ package dns
+	if majorProductVersion >= 11 {
+		// Big Sur, a killall -HUP mDNSResponder doesn't restart the mDNSResponderHelper but killall -TERM of both, does`
+		dlog.Debug(c, "Flushing DNS")
+		devNullCmd(c, "killall", "-HUP", "mDNSResponder")
```
Is this supposed to be "killall", "-TERM" based on the comment? or am I misunderstanding

thallgren replied:
Oh, that comment is wrong. Need to fix.

josecv commented on pkg/client/daemon/dns/flush_darwin.go:
```diff
@@ -2,15 +2,58 @@ package dns
-	_ = dexec.CommandContext(c, "dscacheutil", "-flushcache").Run()
+	dlog.Debug(c, "Flushing DNS")
+	if majorProductVersion >= 11 {
+		// Big Sur, a killall -HUP mDNSResponder doesn't restart the mDNSResponderHelper but killall -TERM of both, does`
+		dlog.Debug(c, "Flushing DNS")
+		devNullCmd(c, "killall", "-HUP", "mDNSResponder")
```
wait i'm confused, maybe i'm reading this wrong but it sounds like the comment doesn't match the code -- should we be doing `killall -TERM` instead of `-HUP` here?

thallgren replied:
The comment is wrong. Will fix.

donnyyung commented on pkg/client/connector/userd_grpc/service.go:
```diff
@@ -77,39 +77,54 @@ func (s *service) Version(_ context.Context, _ *empty.Empty) (*common.VersionInf
 		return result, nil
 	}
-	c = s.callCtx(c, "CreateIntercept")
+	dlog.Debug(c, "called")
```
I think this is an extra "called" log?

Commit:
Changes in response to code-review.

- Remove duplicated debug log of "Flush DNS"
- Fix incorrect comment regarding killall -TERM
- Remove duplicated "called" log for CreateIntercept

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
